### ISSUE TITLE
What is the biggest advantage of separating writes/reads (ex. requests/responses) compared to handling them directly?

### ISSUE THREAD

**Issue by lmatteis** [staltz/ama#22]
One big question that comes up often when discussing with folks about Cycle.js is the separation of writes/reads via drivers. A familiar example of Cycle's reactive loop:

```js
function main(sources) {
  const response$ = sources.HTTP
    .select('foo')
    .flatten()
    .map(response => response);
  const request$ = xs.of({
    url: `https://api.github.com/users/foo`,
    category: 'foo',
  });
  const sinks = {
    HTTP: request$
  };
  return sinks;
}
```

When I show colleagues/friends this example they're immediately struck by the idea of having the request happen *after* the response in the program's natural reading flow. To me this makes programs more declarative, but [others are willing to argue](https://medium.com/@JamesDiGioia/mergemap-does-not-cause-a-side-effect-d016bc997efe#.c6xjqa80e) that reactive streams are declarative even when used in this way:

```js
const fetchUserEpic = (action$, store, { getJSON }) =>
  action$.ofType('FETCH_USER')
    .mergeMap(() =>
      getJSON(`/api/users/${payload}`)
        .map(response => ({
          type: 'FETCH_USER_FULFILLED',
          payload: response
        }))
    );
```

To me however, even though this function is referentially transparent, the chain of observables will still cause a side-effect midway. The fact that it happens midway is what makes it more imperative in my opinion.

Apart from this declarative vs imperative discussion, I was wondering what do you think is the biggest advantage of not mixing flows that cause writes/reads, but rather separate them using drivers. 

In other words: what are the pros and cons of the two snippets of code above?

**Comment by staltz** [comment]
I dwelled into the details of this issue in this talk at GOTO https://www.youtube.com/watch?v=Tkjg179M-Nc

Basically there is a pure way of handling effects that makes code readable in the style of one line after the other. This way is called monadic I/O. It would feel more natural for your colleagues for example. What we use in cycle.js is called stream I/O. Both of these different types of pure I/O are explained in the talk. I also mentioned the talk that stream I/O is not as good as monadic I/O for question and answer style of I/O. HTTP requests and responses are definitely a question and answer style of I/O.

So, to be honest monadic I/O (pure) or imperative programming (impure) are better than stream I/O for these types of I/O, and this is even more visible in case you need to make multiple chained requests and responses. Also in the talk I mention how it is possible to mix stream I/O and monadic I/O and Elm does this.

That said, here are some comments that are outside of what I said in the talk.

(1) in the first code snippet, people may have the impression that the response happens before the request. They need to understand, however, that we are using an entirely different paradigm. In Cycle.js, we are not describing a sequence of actions to do one after the other. This is precisely the idea of imperative programming (or monadic I/O). Instead our paradigm is to describe of pipeline of dataflow. The way our (declarative) description is ordered has nothing to do with the order of computation steps. I feel like more and more we need to understand cycle.js from the perspective of a diagram. In the future I imagine developers using the diagram dev tool side-by-side with a code text editor. If you would have shown a diagram of that first code snippet, maybe your colleagues would have understood that much better. The diagram is just another type of description of the same system that is described also in the code snippet.

(2) About the medium post. mergeMap indeed causes a side effect, and so does getJSON. He mentioned

> mergeMap does not cause a side effect. It creates a new stream that generates and observes other streams. Those other streams, when observed themselves, cause side effects, but those streams aren’t generated or observed or produce side effects until the main stream is observed itself. All of the side effects are generated as a result of the side effect of consuming the stream in the first place.

This just means that the side effect only occurs in the presence of a subscription. However in redux-observable with that mergeMap, the side effect occurs in the operator. In contrast, in cycle.js, the the side effect occurs in the subscription observer. Which leads me to this following point:

> Apart from this declarative vs imperative discussion, I was wondering what do you think is the biggest advantage of not mixing flows that cause writes/reads, but rather separate them using drivers.

The advantage of doing side effects only in subscribe means that the whole pipeline of operators is pure and and therefore more easily testable without invasive mocking. This also means more predictable code/behavior. Plus, it allows you to compose the logic before giving it to drivers. For instance imagine intercepting the HTTP sink and replacing the URL of all of the requests, before they are given to the driver. In cycle.js can do this is easily by composing streams together. I believe it would be a bit harder to do that in redux-observable, and it would probably involve mocking or changing existing code. There's a big difference between composing an existing code and changing existing code. When you are composing, the existing code could come from anywhere, even libraries. But when you are changing existing code, you have to have ownership of that code.

In essence, in redux-observable, the operator pipeline contains both logic and effects. In cycle.js, the operator pipe link contains only logic, and effects are only in the subscription observer.

Does this answer your questions? Feel free to ask a bit more if you want.

**Comment by lmatteis** [comment]
If you don't think request/response models don't fit well the Cycle.js architecture, how do you handle HTTP requests in Cycle?

**Comment by staltz** [comment]
How do I handle? With the Cycle.js architecture. :)
I never said Cycle cannot handle those cases, or that those cases are very bad. I just said monadic I/O or imperative programming feels more natural for effects of the type "make a question, wait for the answer, make another question, etc". But Cycle can handle those for sure.

**Comment by marcusradell** [comment]
Wow, great talk! This was just what I needed!

**Comment by DavidMihola** [comment]
I don't want to hijack this issue at all - but since this @lmatteis's question reminded me of some other thing that I was going to ask you about anyway, I'll go ahead and post his here. I hope that's OK!

I have the feeling that Rx (my experience is with RxJava mostly, but I think this applies to other implementations as well) tends to "complect" (see Rich Hickey's talk "Simple made easy" https://www.infoq.com/presentations/Simple-Made-Easy) the _interest in_ some piece/stream of data with the actual _triggering of the work_ that needs to be done to produce that data.

That is, by subscribing to an `Observable` you BOTH register for updates AND trigger, e. g., a network request. In some cases this is exactly what you want to do, but in almost every project I am working on there will be some situation where this model breaks, and some events have to be "manually" transferred into some other streams: For example, I would trigger (i. e. _subscribe to_) a `getAvailableChannels` request in some long lived view. Later on, from a different view, I would trigger a purchase that unlocks additional channels, which the purchase response will contain. I now need to break open the neat "pipe" that would normally only send these new channels to the second `Subscriber` and splice those new channels into the `Observable` that the first view subscribed to. (Typically with `doOnNext` and a `Subject` that is `merge`d with the original response).

Of course, this problem disappears if I just send out requests at one end and listen to responses at the other end - in this way, actual network requests will be triggered by sending out request objects (and _not_ by someone subscribing to a cold `Observable`) and, conversely new data can be delivered to anyone who is interested (and _not_ only to whomever triggered the request).

So, @staltz, do you agree that this is an advantage of the cycle architecture and/or was this even a design goal that you had in mind?

**Comment by staltz** [comment]
> by subscribing to an Observable you BOTH register for updates AND trigger

I can see how it seems like complecting. However I'd like to argue that Observables are simple (in Rich Hickey's sense) primitives that allow you to achieve anything. To argue for that, I need to introduce two concepts: `IObservable` and `Observable`.

[`IObservable`](https://msdn.microsoft.com/en-us/library/dd990377(v=vs.110).aspx) is originally from Rx in C#, and represents just the interface `{subscribe: function (observer): Subscription}`. In essence this represents "something that can be subscribed", and therefore `IObservable` deals **only** with the *interest in some data*. IObservable says nothing about when the underlying execution is triggered.

`Observable implements IObservable` is the typical case of a "cold" Observable that both registers for updates and triggers execution. 

Note also that `Subject implements IObservable` and Subjects typically do not trigger execution when subscribed to.

So, to answer your question, from the perspective of modelling an API, only the IObservable matters to consumers of the API. A consumer/observer can never tell whether the emissions it receives are result of immediate execution or cached results or whatnot. And more importantly, a consumer shouldn't care about it. In an ideal API, the consumer should only be interested in the data from the producer. (Note that the consumer receiving an Observable can apply operators like `retry`, `take(n)` etc to customize the consumption a bit further).

And when it comes to modelling how executions should trigger, you can choose different combinations of Subjects, ConnectableObservables/refCount, higher order flattening (mergeMap,switchMap,concatMap), and `defer`. I feel like the utility of higher order flattening is underused. They are highly useful for many kinds of delay/delegation of execution triggers. E.g. if you have `getAvailableChannels` returning an Observable, you can subscribe to `clickObservable.switchMap(getAvailableChannels)` instead of subscribing to `getAvailableChannels` directly. Replace `clickObservable` with any other convenient observable dictating the execution trigger. Add `retry` to handle errors or `shareReplay` to cache results, etc.

Does this answer?

**Comment by DavidMihola** [comment]
Thank you - that's very interesting!

You are right, I was probably unfair in saying that Rx as a whole is _complex_. On the other hand, you seem to agree that "cold" `Observables` really do conflate the two actions of registering and triggering - and since these are so prevalent in most RxJava tutorials etc. it has taken me some time and effort to realize that sometimes the two aspects can and should be seperated.

Very often there is no need to do so - like with the neat chains for network requests you mentioned in your first reply. And often, too, there is indeed a way to create longer chains by pushing back (or up?) some events that would otherwise need to be handled in the view layer ("Actually, let's not listen for network changes and then _redo_ our `getUser` request - let's just push the changes up and `flatMap` the `getUser` over them").

But occasionally, like in my example above, I really needed to introduce some `Subject` to push values around. This mixing-in of `Subjects` always felt like a shameful violation of "reactive" principles. And so it was eye opening for me to see an architecture that moved all subscriptions to a special place - and there just used `Subject`s to create the cycle. This seemed to me to be a principled way of using `Subject`s than sprinkling them here and there _ad hoc_.

Anyway, I am not telling you anything new here... thanks again for your reply!

**Comment by staltz** [comment]
> I really needed to introduce some Subject to push values around. This mixing-in of Subjects always felt like a shameful violation of "reactive" principles.

I really understand your concerns, since I'm also not so keen on using Subjects. But they don't inherently make code less reactive (they just make it less functional because Subjects contain state). Subjects are simply the combination of Observer interface with IObservable. They become nasty when people abuse the Observer side of the Subject to imperatively feed in events into its IObservable side. That said, in my opinion the most important part of Subject is that it's the only primitive in Rx which **maintains a list of Observers**. No other piece in Rx does this, and it's essential for sharing executions. Usually this is done **implicitly**, because there is always a Subject underneath a ConnectableObservable (e.g. `share`, `publish` + `refCount`).

About Cycle's architecture, maybe the most important distinction is the way it's functional by putting effects outside the main application. But the Rx techniques I mentioned above are equally reactive, when compared to Cycle.

**Comment by nickbalestra** [comment]
So, from this thread and your GOTO talk you refer to the possibility of using monadic I/O within streams I/O (and therefore the cycle.js architecture). In order to achieve this would you rely on a specific "monadic" driver?

Totally agree with
> In the future I imagine developers using the diagram dev tool side-by-side with a code text editor.

I just gave a talk on cycle.js on a meetup where I did some live coding ending up showing the diagram, and after some discussion with the attendees it would have been so much better to just demo it the other way around. I'll experiment with this!.

**Comment by staltz** [comment]
> So, from this thread and your GOTO talk you refer to the possibility of using monadic I/O within streams I/O (and therefore the cycle.js architecture). In order to achieve this would you rely on a specific "monadic" driver?

Not a monadic driver, but it would just be a matter of making a(ny) driver support `Stream<Instructions>` as sinks, where `Instructions` is the monadic type. As an example, it could even be `type Instructions = Array<string>` where each string is a command that gets results from the previous command and does something else.

```js
const instruction$ = xs.of(
  ['request cycle.js.org', 'from response, find the 7th link on the page', 'request that', 'send back response to sources.HTTP in category Foo']
);
```

But notice how the last instruction in such monadic sequence would have to be a way of passing the results back to `sources` with a category, or any kind of technique we already use in Cycle.js. This exact thing happens also in Elm tasks. If you want to get the result of the chain of instructions, the result must go back to sources. We have a monad type trapped inside a sink stream, and we can't get anything out from sink stream because they are meant only for writes.

Notice how Elm `Http.send` takes a Msg as the first argument. Msg is like Cycle source, it's something that your app reads from the external world. http://package.elm-lang.org/packages/elm-lang/http/1.0.0/Http#send

```elm
import Http

type Msg = Click | NewBook (Result Http.Error String)

update : Msg -> Model -> Model
update msg model =
  case msg of
    Click ->
      ( model, getWarAndPeace )

    NewBook (Ok book) ->
      ...

    NewBook (Err _) ->
      ...

getWarAndPeace : Cmd Msg
getWarAndPeace =
  Http.send NewBook <|
    Http.getString "https://example.com/books/war-and-peace.md
```

So, we *could* have monads trapped in sink streams, but inevitably, the results would have to be fed back to sources in the same way we already do things. Having monadic IO in Cycle would help in very few cases. The most obvious one is to make chained requests, but just the fact of doing chained requests on the frontend is a code smell because there should probably be a backend doing this aggregation. Or maybe it is a problem with REST API design. Anyway, it seems like even though we have a disadvantage not using monadic effects, they wouldn't bring us much benefit because Cycle is meant to solve UIs and frontend.

**Comment by SteveALee** [comment]
Just my 2c's. From a background of event driven code there are advantages to having a single point of events into the module / component. If you take a systems view of your component then there is a single then that's one place events flow in and one place they flow out. If you add a Ajax request and handling inline then you've added to 2 more flows and so double the complexity. More complexity means more scope for bugs and more places to monitor or look for issues.

Another perspective is the component sends a trigger into it's external environment and a result may come back or may not. In fact we should always set a time-out to stay in control of no response happening. When the result or timeout event happen we need to tie the new input flow to our request and we do that one of 2 ways. a) by the code flow and an implicit or explicit async wait. b) provide state to connect them. The state may be and id that we send and comes back in the response as per cycle HTTP categories, or by implementing a finite state machine so a different handler is used for the event.

In my old days of Windows 3 development we did the same. You provide one big case statement for all events (messages) and added custom events for similar async happenings. Before that I was doing the same with embedded real time operating systems - a single process input point and exit point for messages and signals

Thus for me it just feels right :)

Even though cycle splits the input flows into different  "channels" through the drivers and associated stream logic and we don't have one big case statement handling all, the external flows are the same for events that may or may not come back async for an event we trigger (eg a HTTP request).

**Comment by SteveALee** [comment]
Oh a final angle occurred to me, based on RTOS work. When your system is idle it is usually sitting in an input loop waiting for an event which it then dispatches. This may be hidden in a driver or the OS call the driver uses, but it's there. If you call GetJSON which only returns after the response or a timeout then you have added another loop. This may or may not behave the same way as the main loop. It's much better to minimise these places, especially when you need to monitor what's going on.
